go-controller/pkg/ovn/namespace.go

```diff
@@ -135,112 +134,137 @@ func (oc *Controller) multicastUpdateNamespace(ns *kapi.Namespace) {
+	nsInfo.multicastEnabled = enabled
 }
 
 // Cleans up the multicast policy for this namespace if multicast was
 // previously allowed.
-func (oc *Controller) multicastDeleteNamespace(ns *kapi.Namespace) {
-	if oc.multicastEnabled[ns.Name] {
+func (oc *Controller) multicastDeleteNamespace(ns *kapi.Namespace, nsInfo *namespaceInfo) {
+	if nsInfo.multicastEnabled {
 		if err := deleteMulticastAllowPolicy(ns.Name); err != nil {
 			klog.Errorf(err.Error())
 		}
 	}
-	delete(oc.multicastEnabled, ns.Name)
+	nsInfo.multicastEnabled = false
 }
 
 // AddNamespace creates corresponding addressset in ovn db
 func (oc *Controller) AddNamespace(ns *kapi.Namespace) {
 	klog.V(5).Infof("Adding namespace: %s", ns.Name)
-	oc.namespaceMutexMutex.Lock()
-	if oc.namespaceMutex[ns.Name] == nil {
-		oc.namespaceMutex[ns.Name] = &sync.Mutex{}
-	}
-
-	// A big fat lock per namespace to prevent race conditions
-	// with namespace resources like address sets and deny acls.
-	oc.namespaceMutex[ns.Name].Lock()
-	defer oc.namespaceMutex[ns.Name].Unlock()
-	oc.namespaceMutexMutex.Unlock()
-
-	oc.namespaceAddressSet[ns.Name] = make(map[string]string)
+	nsInfo := oc.createNamespaceLocked(ns.Name)
+	defer nsInfo.Unlock()
 
 	// Get all the pods in the namespace and append their IP to the
 	// address_set
 	existingPods, err := oc.watchFactory.GetPods(ns.Name)
+	addresses := make([]string, 0)
 	if err != nil {
 		klog.Errorf("Failed to get all the pods (%v)", err)
 	} else {
 		for _, pod := range existingPods {
 			if pod.Status.PodIP != "" {
 				portName := podLogicalPortName(pod)
-				oc.namespaceAddressSet[ns.Name][pod.Status.PodIP] = portName
+				nsInfo.addressSet[pod.Status.PodIP] = portName
+				addresses = append(addresses, pod.Status.PodIP)
 			}
 		}
 	}
 
-	addresses := make([]string, 0)
-	for address := range oc.namespaceAddressSet[ns.Name] {
-		addresses = append(addresses, address)
-	}
-
 	// Create an address_set for the namespace.  All the pods' IP address
 	// in the namespace will be added to the address_set
 	createAddressSet(ns.Name, hashedAddressSet(ns.Name), addresses)
 
-	oc.namespacePolicies[ns.Name] = make(map[string]*namespacePolicy)
-	oc.multicastUpdateNamespace(ns)
+	oc.multicastUpdateNamespace(ns, nsInfo)
 }
 
 func (oc *Controller) updateNamespace(old, newer *kapi.Namespace) {
-	klog.V(5).Infof("Updating namespace: old %s new %s", old.Name, newer.Name)
+	klog.V(5).Infof("Updating namespace: %s", old.Name)
 
-	// A big fat lock per namespace to prevent race conditions
-	// with namespace resources like address sets and deny acls.
-	oc.namespaceMutex[newer.Name].Lock()
-	defer oc.namespaceMutex[newer.Name].Unlock()
+	nsInfo := oc.getNamespaceLocked(old.Name)
+	if nsInfo == nil {
+		klog.Warningf("Update event for unknown namespace %q", old.Name)
+		return
+	}
+	defer nsInfo.Unlock()
 
-	oc.multicastUpdateNamespace(newer)
+	oc.multicastUpdateNamespace(newer, nsInfo)
 }
 
 func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
```

Comment:
A question: If namespace A is deleted then re-created, do we have any sort of guarantee that those events are processed in-order? I *think* so, since the client-go informers handle that correctly.

Reply:
I think we are guaranteed to get them in order, how we process them with the current queuing system ensures the events of the same type go to the same queue (in order).  @dave-tucker is looking into replacing our current custom queues with the k8s workqueues, so he may have a more insight here

Reply:
Yes, that's guaranteed